### PR TITLE
Collection filter

### PR DESCRIPTION
**Pull request by rowanwins**
This PR adds support for more advanced search filters using the `/queryables` endpoint to retrieve fields which can be queried, and then generating POST requests using cql2 based on the users inputs.

https://user-images.githubusercontent.com/6735870/170149205-dde3c479-e03f-401a-8e92-86bc5a7dfc5a.mp4

Still needs a bit of tidying but coming along.
cc @m-mohr

### CODE REVIEW

**Comment by m-mohr** [comment]
Thank you @rowanwins. Just want to let you know that I've seen that, but can't respond right now due to a conference this week. I'll get back to you in roughly a week or so.

**Comment by rowanwins** [comment]
The latest commits
- Add support for using a field multiple times (eg you want to see a cloud cover filter with upper and lower bounds)
- Shifts to using a dropdown instead of a select field because vue-bootstrap was having some strange issues

**Comment by m-mohr** [comment]
@rowanwins I've added eslint rules to the v3 branch, which then required a couple of changes to the files. I've resolved merge conflicts in both PRs, but I think you still need to run `npm run lint` and check both PRs for compliance.

File: src/models/Queryable.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+  async getDefinitionFromReference () {
+    const response = await fetch(this._rawJson.$ref)
+    if (!response.ok) return
+    const data = await response.json()
+    
+    const uri = new URI(this._rawJson.$ref)
+    const hash = uri.hash()
+    const hashComponents = hash.replace('#/', '').split('/').join('.')
+    return Object.assign(this._rawJson, _get(data, hashComponents))
```
Comment (m-mohr) [inline src/models/Queryable.js:66]: As far as I can see this is used only once, but adds (for the functionality given) a relatively large additional dependency (the source has 900 lines). 

If I understand the docs correctly, we can simply do/add the following instead:
```js
function _get(object, path) {
  if (object === null || typeof object !== 'object') {
    return;
  }
  object = object[path[0]];
  if (typeof object !== 'undefined' && path.length >1) {
    return _get(object, path.slice(1));
  }
  return object;
}
```

You'd need to remove the `join('.')` in the line above.

File: package.json [hunk 2]
```diff
@@ -32,12 +32,14 @@
...
     "@radiantearth/stac-fields": "1.0.0-beta.13",
     "@radiantearth/stac-migrate": "~1.0.3",
+    "@turf/bbox-polygon": "^6.5.0",
     "axios": "^0.21.4",
     "bootstrap-vue": "^2.21.2",
     "bs58": "^4.0.1",
     "commonmark": "^0.29.3",
     "core-js": "^3.6.5",
     "leaflet": "~1.7.1",
+    "lodash.get": "^4.4.2",
```
Comment (m-mohr) [inline package.json:42]: I'd propose to remove this, see https://github.com/radiantearth/stac-browser/pull/157/files#r889715354

File: package.json [hunk 3]
```diff
@@ -32,12 +32,14 @@
     "@lweller/leaflet-areaselect": "^1.3.0",
     "@radiantearth/stac-fields": "1.0.0-beta.13",
     "@radiantearth/stac-migrate": "~1.0.3",
+    "@turf/bbox-polygon": "^6.5.0",
```
Comment (m-mohr) [inline package.json:35]: Is this used anywhere? I can't find this being used so it might be removed.

File: src/components/ItemFilter.vue [hunk 4]
```diff
@@ -40,26 +40,62 @@
         />
       </b-form-group>
 
+      <div class="additionalFilters">
+        <b-form-group label="Select additional field filters" label-for="availableFields" description="Fields advertised by the /queryables endpoint">
```
Comment (m-mohr) [inline src/components/ItemFilter.vue:44]: I'd remove the description (or make it less technical). /queryables is a technical details normal users don't care about.

File: src/components/ItemFilter.vue [hunk 5]
```diff
@@ -40,26 +40,62 @@
...
+          <b-dropdown size="sm" text="Add field filter" block variant="primary" class="m-2" menu-class="w-100">
+            <b-dropdown-item v-for="option in fieldFilterOptions" :key="option.text" @click="additionalFieldSelected(option)">{{ option.text }}</b-dropdown-item>
```
Comment (m-mohr) [inline src/components/ItemFilter.vue:46]: I've tested this against MS PC (prod and staging), but the lift of filters is empty for me?

File: src/components/ItemFilter.vue [hunk 6]
```diff
@@ -40,26 +40,62 @@
         />
       </b-form-group>
 
+      <div class="additionalFilters">
+        <b-form-group label="Select additional field filters" label-for="availableFields" description="Fields advertised by the /queryables endpoint">
+          <b-dropdown size="sm" text="Add field filter" block variant="primary" class="m-2" menu-class="w-100">
```
Comment (m-mohr) [inline src/components/ItemFilter.vue:45]:
```suggestion
          <b-dropdown size="sm" text="Add filter" block variant="primary" class="m-2" menu-class="w-100">
```

File: src/components/ItemFilter.vue [hunk 7]
```diff
@@ -121,33 +168,53 @@ export default {
...
-        sortby: null
+        sortby: null,
+        advancedFilters: null
       };
     },
     onSubmit() {
       if (this.sort) {
         this.filters.sortby = this.formatSort();
       }
+      this.filters.advancedFilters = this.itemSearch.getAsCql2Json();
```
Comment (m-mohr) [inline src/components/ItemFilter.vue:212]: In global search I'm getting for this line:

> TypeError: this.itemSearch is null

File: src/components/ItemFilter.vue [hunk 8]
```diff
@@ -88,7 +124,8 @@ export default {
   },
   data() {
     return {
-      sortOrder: 1,
+      itemSearch: null,
+      sortOrder: 'asc',
```
Comment (m-mohr) [inline src/components/ItemFilter.vue:128]: Changing the sortOrder from int to string seems to not be fully working (what was the reason for it?). I'm getting:

> Invalid prop: type check failed for prop "value". Expected String with value "0", got Number with value 0.

File: src/models/Queryable.js (hunk 1 above)
Comment (rowanwins) [reply src/models/Queryable.js]: Happy to test the proposed.
I generally look at bundleophobia before adding a package to check it's size contributions, [in this case it's 2kb](https://bundlephobia.com/package/lodash.get@4.4.2), which isn't terribly large, but obviously larger than your snippet so if we can trim that down why not :)

File: src/components/ItemFilter.vue (hunk 5 above)
Comment (rowanwins) [reply src/components/ItemFilter.vue:46]: I haven't done the global search side of things yet, I've been focussed on doing things at the collection level and getting my head around the app architecture, but I'll look into the global search side now.

File: src/components/ItemFilter.vue (hunk 7 above)
Comment (rowanwins) [reply src/components/ItemFilter.vue:212]: See comment above, moving onto global search now.

File: src/components/ItemFilter.vue (hunk 8 above)
Comment (rowanwins) [reply src/components/ItemFilter.vue:128]: Yeah this was/is a fiddily one. Basically what I was experimenting with was trying to get things more aligned with the STAC spec out of the box - ATM the filter has to pass things through a formatter to put it into a format compliant with the API's. 
 https://github.com/radiantearth/stac-browser/blob/e780a3a1625ba0684fabfef2d4fcca252a9228a1/src/components/ItemFilter.vue#L195-L202
I think I also ran into issues with one of the API's (MSC PC probably) not respecting the `+` or `-` values, and only accepting `asc` or `desc`.

When I implemented the changes I wasn't thinking about sorting of [catalog](https://github.com/radiantearth/stac-browser/blob/v3/src/components/Catalogs.vue#L48-L53) or [items](https://github.com/radiantearth/stac-browser/blob/v3/src/components/Items.vue#L94-L99) lists, which happen outside of an API.

So I'll have a think about how to reconcile this.

Comment (rowanwins) [reply src/components/ItemFilter.vue]: Have fallen back to the old approach.

**Comment by rowanwins** [comment]
Alright I think this is getting pretty close now @m-mohr .

**Comment by m-mohr** [comment]
Thank you. I've merged this to get the huge amount of changes in. Now it's easier to write specific issues and work on smaller tasks to evolve from there.

I'll apply some style changes that I planned to do anyway for the filters.

**Pull request merged**